src/function/memo.rs

```diff
@@ -189,7 +188,10 @@ impl<V> Memo<V> {
                     // We hit a cycle blocking on the cycle head; this means it's in
                     // our own active query stack and we are responsible to resolve the
                     // cycle, so go ahead and return the provisional memo.
-                    return false;
+                    tracing::debug!(
+                        "Waiting for {head_index:?} results in a cycle, return {database_key_index:?} once all other cycle heads completed to allow the outer cycle to make progress."
+                    );
+                    hit_cycle = true;
```

Comment:
I changed this to early-return in https://github.com/salsa-rs/salsa/pull/861 but I think that was incorrect. Not early-returning fixes the ty panic for `kopf`. I might spend some more time to come up with a test case for it but I might also just leave it at it (I already spent another week on this :s)